#summary How to use ZINBA
<wiki:toc max_depth="2" />

= Using ZINBA =
A <a href="http://www.unc.edu/~nur2/zinbaweb/test_data.tgz">test dataset</a> has been created that corresponds to the example code given in each function used below. 

=Necessary file downloads=
Before using ZINBA, you need two files before you can start your analysis, given below.   

==1. Genome Build==
The build of the genome you are using in .2bit format, downloadable below.  Also downloadable from the from the <a href=" http://hgdownload.cse.ucsc.edu/downloads.html">UCSC genome browser</a>. NOTE: YOUR DOWNLOADED BUILD MUST BE THE SAME BUILD THAT YOUR READS WERE MAPPED TO. 
||*Human*||*Mouse*||*Drosophila*||*C. Elegans*||
||<a href="http://hgdownload.cse.ucsc.edu/goldenPath/hg18/bigZips/hg18.2bit"  >hg18</a>||<a href="http://hgdownload.cse.ucsc.edu/goldenPath/mm8/bigZips/mm8.2bit"  >mm8</a>||<a href="http://www.bios.unc.edu/~nrashid/dm3.2bit" >dm3</a>||<a href="http://www.bios.unc.edu/~nrashid/ce4.2bit">ce4</a>||
||<a href="http://hgdownload.cse.ucsc.edu/goldenPath/hg18/bigZips/hg19.2bit"  >hg19</a>||<a href="http://hgdownload.cse.ucsc.edu/goldenPath/mm9/bigZips/mm9.2bit"  >mm9</a>|| ||<a href="http://www.bios.unc.edu/~nrashid/ce6.2bit">ce6</a>|| 

	
==2. Mappability File==
A mappability folder corresponding to your <u>sequencing read length</u> and <u>genome</u> (links to compressed files are below).  These files were generated using code from <a href="www.gersteinlab.org/proj/PeakSeq">Peakseq </a>, developed by the Gerstein Lab. This file is used to derive final mappability score in ZINBA.

||*Human*||*Mouse*||*Drosophila*||*C. Elegans*||
||<a href="http://www.bios.unc.edu/~nrashid/map36.tgz"  >hg18 36bp</a><br><a href="http://www.bios.unc.edu/~nrashid/map50.tgz"  >hg18 50bp</a> || ||<a href="http://www.bios.unc.edu/~nrashid/map36_dm3.tgz"  >dm3 36bp</a>||<a href="http://www.bios.unc.edu/~nrashid/map36_ce4.tgz"  >ce4 36bp</a> ||
||<a href="http://www.bios.unc.edu/~nrashid/map36_hg19.tgz"  >hg19  36bp</a><br><a href="http://www.bios.unc.edu/~nrashid/map50_hg19.tgz"  >hg19 50bp</a> ||<a href="http://www.bios.unc.edu/~nrashid/map36_mm9.tgz"  >mm9 36bp</a><br><a href="http://www.bios.unc.edu/~nrashid/map50_mm9.tgz"  >mm9 50bp</a> || <a href="http://www.unc.edu/~nur2/map50_dm3.tgz"  >dm3 50bp</a> ||<a href="http://www.bios.unc.edu/~nrashid/map36_ce6.tgz"  >ce6 36bp</a>|| 


=Preparing your files=
  # Unpack your downloaded mappability folder by changing to the directory it is located in and running the command below, where ## is the specific name/number corresponding to the file 
{{{
wget http://www.bios.unc.edu/~nrashid/map##.tgz
tar -xzvf map##.tar.gz
}}} 
  # Enter R and feed the unpacked mappability folder to the generateAlignability() function to generate your alignability directory. You do not need to compute this again unless you are running samples mapped with different alignability thresholds (athresh) or whose sequencing library has a different average fragment length (extension)
{{{  
generateAlignability(
   mapdir=,      #mappability directory from unpacked mappability files
   outdir=,      #directory for processed files, used later in analysis
   athresh=,     #number of hits per read allowed during mapping process
   extension=,   #average fragment library length
   twoBitFile=,  #path to downloaded genome build file in .2bit format
)
}}} 
  # Run the basealigncount function to generate the basecount file needed to obtain exact peak boundaries through peak refinement.  If peak refinement is not desired then this step can be skipped.  Otherwise, this should be generated for each set of sample reads you are analyzing.  In most cases the unrefined estimates should be sufficient.
{{{
basealigncount(
   inputfile=,  #mapped sample reads
   outputfile=, # output path
   extension=,  #average fragment library length 
   filetype=,   #either "bed", "bowtie", or "tagAlign"
   twoBitFile=, #path to downloaded genome build file in .2bit format
)
}}}


=Analyze your Data=
==Input Formats==
Three input formats are allowed for ZINBA:

*BED*:  The first [http://genome.ucsc.edu/FAQ/FAQformat.html#format1 6 fields] of the BED format
  * Chromosome
  * Start
  * Stop
  * Name (can be any string/character)
  * Score (can be set to 0 if N/A)
  * Strand (can be set to + if unknown)

*tagAlign*:  Similar to the above BED format, except "Seq" replaces "Name".  This format is sometimes found on the UCSC genome browser.
  * Chromosome
  * Start
  * Stop
  * Seq (sequence of mapped read)
  * Score (can be set to 0)
  * Strand (can be set to + if unknown)

*bowtie*  The [http://bowtie-bio.sourceforge.net/manual.shtml#default-bowtie-output default output] from the Bowtie mapping software (.hits file)

==ZINBA pipeline function==
Usage for the ZINBA pipeline function
{{{
zinba(
  refinepeaks=,   #refine peaks?  1 for yes, 0 for no
  seq=,           #path to mapped experimental reads
  input=,         #path to mapped input reads if available (default is "none")
  filetype=,      #either 'bed', 'bowtie', or 'tagAlign'
  threshold=,     #FDR threshold, default is 0.05 
  align=,         #path to alignability directory
  numProc=,       #number of CPUs to use, must be less than max available (default 1)
  twoBit=,        #path to genome build in .2bit format
  outfile=,       #prefix for outputted files
  extension=,     #average fragment library length (size selected)
  
  ###################
  OPTIONAL PARAMETERS
  ###################

  basecountfile=, #path to basecount file if refinepeaks is 1
  broad=,         #broad setting, TRUE or FALSE (default)
  printFullOut=,  #print original data with enrichment estimates, 1  for yes (more space required), 0 for no (default)
  interaction=,   #whether or not to considering interaction during model selection, TRUE (default) or FALSE
  mode=,          #either "peaks" for peak calling (default) or "CNV" for calling likely amplified CNV regions for reads in "seq" (input reads are best)
  FDR=            #either TRUE (default) or FALSE.  If false, then uses posterior probability to threshold peaks using 1-threshold 
)
}}}

This function is optimized for speed and for ease of use, utilizing only input if it is available.  A more flexible function run.zinba is available where more parameters can be specified. 

For ENCODE reporting purposes, the zinba() convenience function can convert posterior probability to q-values for FDR thresholding if one sets FDR=TRUE (more liberal results than the posterior probability-based results in ZINBA manuscript).    
    
==FAIRE-seq example==
FAIRE data example: sequencing reads are 36 bp in length, and correspond to a fragment library length of 134 base pairs. The mappability file is unpacked, unloading the map36/ folder with the mappability files contained in it. We create the alignability directory at align_athresh4_extension134/, where our reads were mapped with an alignability threshold of matching up to 4 regions in the genome (otherwise they were filtered out). Download the necessary files, unpack your mappability directory, and create a directory to place your alignability files. Open a terminal and run the following, corresponding to the test unpacked dataset (a data/ directory will be created wherever you unpacked it). 

{{{
#open a terminal and download the necessary files and test dataset.  
#on the mac, use curl -O instead of wget
wget http://hgdownload.cse.ucsc.edu/goldenPath/hg18/bigZips/hg18.2bit
wget http://www.bios.unc.edu/~nrashid/map36.tgz
wget http://www.unc.edu/~nur2/zinbaweb/test_data.tgz

#unpack your mappability folder and test datasets
tar -xzvf map36.tgz
tar -xzvf test_data.tgz

#create your directory to hold your alignability files.  You can do this anyway you please, terminal command is below
mkdir align_athresh4_extension134
}}}

Enter R and generate your alignability directory in addition to your basecount file 
{{{
library(zinba)
generateAlignability(
   mapdir='map36/',
   outdir='align_athresh4_extension134/',
   athresh=4,
   extension=134,
   twoBitFile='hg18.2bit'
)   

basealigncount(
   inputfile='data/faireGM12878rep1chr22.taf',
   outputfile='data/faireGM12878rep1chr22.basecount', 
   extension=134, 
   filetype='tagAlign',
   twoBitFile='hg18.2bit'
)

}}}

Now, run your zinba analysis. Here we start from our mapped sample reads and our prepared files, build the datasets needed for analysis, run the mixture regression model with the model selection specified, and then run a peak boundary refinement to capture exact peak boundaries. This function is our pipeline function, tying together the smaller functions that work on different parts of the analysis. Because input control was not available for this data, this function only considers the covariates GC content, mappability, and our estimate for local background to model the data, and utilizes the model selection procedure to select the best model for each component. 

{{{
zinba(
	align='align_athresh4_extension134/',
	numProc=4,
	seq='data/faireGM12878rep1chr22.taf',
	basecountfile='data/faireGM12878rep1chr22.basecount',
	filetype="tagAlign",
	outfile="data/faire",
	twoBit="hg18.2bit",
	extension=134,
	printFullOut=1,
	refinepeaks=1,
	broad=F,
	input="none"
)

    
}}}
==ChIP-seq example==
ChIP-seq data example: sequencing reads are 36 bp in length, and correspond to a fragment library length of 200 base pairs. The mappability file is unpacked, unloading the map36/ folder with the mappability files contained in it. We create the alignability directory at align_athresh1_extension200/, where our reads were mapped with an alignability threshold of matching up to 1 region in the genome (otherwise they were filtered out). Download the necessary files, unpack your mappability directory, and create a directory to place your alignability files. Open a terminal and run the following, corresponding to the unpacked test dataset (a data/ directory will be created wherever you unpacked it). 

{{{
#open a terminal and download the necessary files and test dataset.  
#on the mac, use curl -O instead of wget
wget http://hgdownload.cse.ucsc.edu/goldenPath/hg18/bigZips/hg18.2bit
wget http://www.bios.unc.edu/~nrashid/map36.tgz
wget http://www.unc.edu/~nur2/zinbaweb/test_data.tgz

#unpack your mappability folder and test datasets
tar -xzvf map36.tgz
tar -xzvf test_data.tgz

#create your directory to hold your alignability files.  You can do this anyway you please, terminal command is below
mkdir align_athresh1_extension200
}}}

Enter R and generate your alignability directory in addition to your basecount file.

{{{
library(zinba)
generateAlignability(
   mapdir='map36/',
   outdir='align_athresh1_extension200/', 
   athresh=1,  
   extension=200,  
   twoBitFile='hg18.2bit'
)   

basealigncount(
   inputfile='data/ctcfGM12878rep3chr22.taf', 
   outputfile='data/ctcfGM12878rep3chr22.basecount', 
   extension=200, 
   filetype='tagAlign',
   twoBitFile='hg18.2bit' )
}}}

Now, run your zinba analysis. Here we start from our mapped sample reads and our prepared files, build the datasets needed for analysis, run the mixture regression model considering only input control, and then run a peak boundary refinement to capture exact peak boundaries.  

{{{
zinba(
  align='align_athresh1_extension200/',
  numProc=4,			
  seq='data/ctcfGM12878rep3chr22.taf',
  basecountfile='data/ctcfGM12878rep3chr22.basecount',
  filetype="tagAlign",
  outfile="data/faire",
  twoBit="hg18.2bit",
  extension=200,
  printFullOut=1,
  refinepeaks=1,	
  broad=F,
  input='data/inputGM12878rep3chr22.taf'
)
}}}